File: go/llmapi/llm_router.go
package llmapi

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
)

type LlmRouter struct {
	Provider string
	ApiKey   string
	BaseUrl  string
	client   *http.Client
}

func NewLlmRouter(provider string, apiKey string) *LlmRouter {
	baseUrl := "https://api.openai.com/v1/chat/completions" // Default
	if provider == "anthropic" {
		baseUrl = "https://api.anthropic.com/v1/messages"
	}

	return &LlmRouter{
		Provider: provider,
		ApiKey:   apiKey,
		BaseUrl:  baseUrl,
		client:   &http.Client{},
	}
}

// SendRequest now accepts context.Context to support cancellation
func (r *LlmRouter) SendRequest(ctx context.Context, request *LlmRequest) (*LlmResponse, error) {
	reqBytes, err := json.Marshal(request)
	if err != nil {
		return nil, err
	}

	httpReq, err := http.NewRequestWithContext(ctx, "POST", r.BaseUrl, bytes.NewBuffer(reqBytes))
	if err != nil {
		return nil, err
	}

	if r.Provider == "anthropic" {
		httpReq.Header.Set("x-api-key", r.ApiKey)
		httpReq.Header.Set("anthropic-version", "2023-06-01")
		httpReq.Header.Set("Content-Type", "application/json")
	} else {
		httpReq.Header.Set("Authorization", "Bearer "+r.ApiKey)
		httpReq.Header.Set("Content-Type", "application/json")
	}

	resp, err := r.client.Do(httpReq)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
	    return nil, fmt.Errorf("API request failed with status: %d", resp.StatusCode)
	}

	var llmResponse LlmResponse
	if err := json.NewDecoder(resp.Body).Decode(&llmResponse); err != nil {
		return nil, err
	}

	return &llmResponse, nil
}

func (r *LlmRouter) StreamResponse(ctx context.Context, request *LlmRequest, onChunk func(string)) error {
	return nil
}

func (r *LlmRouter) HandleToolCall(toolCall *ToolCall) string {
	return "{}"
}

File: java/src/main/java/com/harness/LlmRouter.java
package com.harness;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.function.Consumer;

public class LlmRouter {

    private final String provider;
    private final String apiKey;
    private final String baseUrl;
    private final HttpClient httpClient;
    private final ObjectMapper objectMapper;

    public LlmRouter(String provider, String apiKey) {
        this.provider = provider;
        this.apiKey = apiKey;

        if ("anthropic".equalsIgnoreCase(provider)) {
            this.baseUrl = "https://api.anthropic.com/v1/messages";
        } else {
            this.baseUrl = "https://api.openai.com/v1/chat/completions";
        }

        this.httpClient = HttpClient.newHttpClient();
        this.objectMapper = new ObjectMapper();
    }

    public String getProvider() {
        return provider;
    }

    public String getApiKey() {
        return apiKey;
    }

    public LlmApi.LlmResponse sendRequest(LlmApi.LlmRequest request) throws Exception {
        String requestBody = objectMapper.writeValueAsString(request);

        HttpRequest.Builder requestBuilder = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl))
                .POST(HttpRequest.BodyPublishers.ofString(requestBody))
                .header("Content-Type", "application/json");

        if ("anthropic".equalsIgnoreCase(provider)) {
            requestBuilder.header("x-api-key", apiKey);
            requestBuilder.header("anthropic-version", "2023-06-01");
        } else {
            requestBuilder.header("Authorization", "Bearer " + apiKey);
        }

        HttpRequest httpRequest = requestBuilder.build();

        HttpResponse<String> response = httpClient.send(httpRequest, HttpResponse.BodyHandlers.ofString());

        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new RuntimeException("API request failed with status: " + response.statusCode());
        }

        return objectMapper.readValue(response.body(), LlmApi.LlmResponse.class);
    }

    public void streamResponse(LlmApi.LlmRequest request, Consumer<String> onChunk) {
        // Placeholder for SSE parsing logic
    }

    public String handleToolCall(LlmApi.ToolCall toolCall) {
        // Placeholder for local function routing
        return "{}";
    }
}
